### ISSUE TITLE
Batch normalization implemented incorrectly in canned DNN estimators

### ISSUE THREAD

**Issue by lukmanr** [tensorflow/tensorflow#21915]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: all
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.10
- **Python version**: 3.6.2
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
Batch normalization is implemented incorrectly in canned DNN estimators.  In _dnn_logit_fn_builder, batch normalization is applied after the dense layer which includes the activation function.  But batch normalization should be applied to the inputs to a layer, prior to the activation function.  See the original paper by Ioffe, Szegedy: " ... we focus on transforms that consist of an affine transformation followed by anelement-wise nonlinearity:
z = g(Wu + b)
where W and b are learned parameters of the model, and g(·) is the nonlinearity such as sigmoid or ReLU. ...  We add the BN transform immediately before the nonlinearity, by normalizing x = Wu+ b."

Also see the many published reference implementations of batch normalization.

The current implementation is not only incorrect but produces very poor results.

**Comment by facaiy** [comment]
ping @ispirmustafa

**Comment by lukmanr** [comment]
Also one must add the update ops for batch norm vars to the train op, cf. the comments for the batch_normalization in layers.normalization.py.  See PR #21916 .

**Comment by ispirmustafa** [comment]
Hi @lukmanr , agreed that it should be applied before relu. re: update-ops, head handles it by default.

**Comment by hjmus** [comment]
Hi @lukmanr,

Can you provide some evidence that the current ordering of BN and activation hurts the performance of your use case?

There are evidences that the ordering in the original paper is not correct: 
http://redd.it/67gonq, 
https://github.com/cvjena/cnn-models/issues/3
https://github.com/ducha-aiki/caffenet-benchmark/blob/master/batchnorm.md
More here:
https://www.google.com/search?q=batch+norm+before+or+after+activation

So apparently this is a topic up for discussion. In my use case I found the result was better with BN following activation.

**Comment by lukmanr** [comment]
Hi @hjmus,

Thanks for the references on ordering; I'll comment below.  The bigger issue for the implementation in canned DNNs is probably the lack of use of the mean/variance update ops.  

@ispirmustafa I don't see any change to train_ops related to BN in the heads.  I also don't see it in the original commit for the feature. You have to add the mean/variance update ops to the train_op like so:

       update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
       with tf.control_dependencies(update_ops):
              train_op = optimizer.minimize(loss)

Otherwise the mean/variance vars in BN layers never get updated and BN is not really in effect.  This is really easy to miss in the current implementation of layers.batch_normalization, and there have been multiple requests to address this, by adding warnings or make the update ops the default behavior:  #21229, #21595, #14809, #14699.  It looks like the Canned DNN implementation missed this as well.  I incuded a suggested fix in the PR.

Re: ordering, I have only previously implemented BN applied before relu.  I find the discussion online confounds multiple issues, including the best explanation for what BN is really doing, with evolving best practices for CNNs.  Worth noting that we are talking about DNNs not CNNs here.  I did try a side-by-side comparison of a custom DNN, with BN before relu (and update ops properly added) against the same architecture in canned DNNs on [this Kaggle data set](https://www.kaggle.com/c/acquire-valued-shoppers-challenge/data) and got terrible results with the canned DNNs, even after hypertuning, which lead me to investigate BN in canned DNNs.  I can't publish the code at the moment as it has yet to go through Google's open sourcing process.

**Comment by facaiy** [comment]
As said by @ispirmustafa ,  update_ops is handled by Head by default:

https://github.com/tensorflow/tensorflow/blob/8dfb7532f8278e53a86a847ba6aa9c441f7b021b/tensorflow/python/estimator/canned/head.py#L1545-L1550

**Comment by ispirmustafa** [comment]
Thank you @hjmus.
I've chat with Christian, Sergey. They don't have a strong opinion about the right order. So let's make this an optional which doesn't change the current behavior . I'll recommend something on the PR.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ispirmustafa** [comment]
Hi @lukmanr 
Could you please update the progress on your PR #21916 ? 
Thank you

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 77 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tilakrayal** [comment]
@lukmanr,
Could you update the PR https://github.com/tensorflow/tensorflow/pull/21916. Respective PR was closed. Thank you
